node/src/components/linear_chain_sync.rs

```diff
@@ -262,20 +331,60 @@ impl<I: Clone + PartialEq + 'static> LinearChainSync<I> {
             }
             State::SyncingTrustedHash {
                 ref mut linear_chain,
+                ref mut current_block,
                 ..
-            } => linear_chain.pop(),
+            } => match linear_chain.pop() {
+                None => None,
+                Some(block) => {
+                    // Update `current_block` so that we can verify whether result of execution
+                    // matches the expected value.
+                    current_block.replace(block.clone());
+                    Some(block)
+                }
+            },
             State::SyncingDescendants {
                 ref mut linear_chain,
+                ref mut current_block,
                 ..
-            } => linear_chain.pop_front(),
+            } => match linear_chain.pop_front() {
+                None => None,
+                Some(block) => {
+                    // Update `current_block` so that we can verify whether result of execution
+                    // matches the expected value.
+                    current_block.replace(block.clone());
+                    Some(block)
+                }
+            },
         };
 
-        match next_block {
-            None => {
-                // We're done syncing but we have to wait for the execution of all blocks.
-                Effects::new()
+        next_block.map_or(Effects::new(), |block| {
+            fetch_block_deploys(effect_builder, peer, block)
+        })
+    }
+
+    fn fetch_next_block<R, REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        rng: &mut R,
+        block_header: &BlockHeader,
+    ) -> Effects<Event<I>>
+    where
+        I: Send + Copy + 'static,
+        REv: ReactorEventT<I>,
+        R: Rng + CryptoRng + ?Sized,
+    {
+        self.reset_peers(rng);
+        let peer = self.random_peer_unsafe();
+        match self.state {
+            State::SyncingTrustedHash { .. } => {
+                let parent_hash = *block_header.parent_hash();
+                fetch_block_by_hash(effect_builder, peer, parent_hash)
             }
-            Some(block) => fetch_block_deploys(effect_builder, peer, block),
+            State::SyncingDescendants { .. } => {
+                let next_height = block_header.height() + 1;
+                fetch_block_at_height(effect_builder, peer, next_height)
+            }
+            _ => panic!("Tried fetching block when in {:?} state", self.state),
```

Comment:
Maybe explicitly list `State::Done | State::None` — that way the compiler will tell us if we add another state and forget to update this.

Reply:
Done.